### PR TITLE
feat: bringing provider governance to budget & limits

### PR DESCRIPTION
**Pull request by roroghost17**
## Summary

Provider-level governance (budget and rate limit) has been migrated from `config_providers.budget_id / rate_limit_id` into `governance_model_configs` as wildcard rows with `scope='global'`, `model_name='*'`, and `provider=<name>`. This makes `governance_model_configs` the single source of truth for all governance enforcement, eliminating the separate provider-governance enforcement path.

## Changes

- **Database migration** (`migrationMigrateProviderGovernanceToModelConfigs`): Folds existing provider-level budget/rate-limit FK references into new `(global, *, <provider>)` model config rows, reusing the same budget/rate-limit rows. Provider FKs are then nulled out. The migration is idempotent and includes a rollback path.
- **`ModelConfigAllModels = "*"` sentinel**: Introduced as a named constant to represent "all models" in a model config row. The `"*"` sentinel is excluded from catalog-based model name normalization.
- **`collectModelConfigsFor`**: New helper that resolves all applicable model configs for a request across four tiers — exact model+provider, exact model (all providers), all models on this provider (`*:provider`), and all models on all providers (`*:nil`) — deduped by config ID. All budget/rate-limit check and usage-tracking paths now use this helper instead of duplicating two-tier lookup logic.
- **`modelConfigEntityKey`**: New helper that builds a stable entity key for a model config, replacing ad-hoc `fmt.Sprintf` strings scattered across check and update functions.
- **`GetProviderGovernanceModelConfigs`**: New store method that queries the wildcard model config rows backing provider governance, with budget and rate-limit preloads.
- **`DeleteProvider` cleanup**: When a provider is deleted, its associated wildcard model configs (and their owned budget/rate-limit rows) are now cleaned up as part of the transaction.
- **Provider governance HTTP handlers**: `getProviderGovernance`, `updateProviderGovernance`, and `deleteProviderGovernance` are rewritten to operate on wildcard model config rows instead of provider FK columns. The GET endpoint gains a `from_memory` query parameter to serve data from the in-memory governance store.
- **UI**: The model limits table renders `"*"` as `"All Models"`. The model limit sheet exposes an `allowAllOption` on the model selector. RTK Query cache tags are cross-invalidated so that changes to model configs refresh the provider governance view and vice versa.

## Type of change

- [ ] Bug fix
- [x] Feature
- [x] Refactor
- [ ] Documentation
- [ ] Chore/CI

## Affected areas

- [x] Core (Go)
- [x] Transports (HTTP)
- [ ] Providers/Integrations
- [x] Plugins
- [x] UI (React)
- [ ] Docs

## How to test

```sh
# Core/Transports
go test ./framework/configstore/... ./plugins/governance/...

# UI
cd ui
pnpm i
pnpm build
```

1. Start with a deployment that has providers with existing `budget_id` or `rate_limit_id` values. After the migration runs, verify those providers have `budget_id = NULL` and `rate_limit_id = NULL`, and that corresponding `(global, *, <provider>)` rows exist in `governance_model_configs` pointing to the same budget/rate-limit IDs.
2. Re-run the migration and confirm no duplicate wildcard rows are created (idempotency).
3. Make a request through a provider that had governance configured and confirm the budget/rate-limit is still enforced.
4. Use `GET /api/governance/providers` and `GET /api/governance/providers?from_memory=true` and confirm both return the expected provider governance data.
5. Use `PUT` and `DELETE` on `/api/governance/providers/{name}` and confirm the wildcard model config rows are created, updated, or removed accordingly, and that the Model Limits UI reflects the change without a manual refresh.
6. Delete a provider and confirm its wildcard model config and owned budget/rate-limit rows are removed.

## Breaking changes

- [x] Yes
- [ ] No

Provider governance is no longer stored in `config_providers.budget_id / rate_limit_id`. Any code or tooling that reads governance directly from the providers table will no longer find it there. All governance enforcement and management must go through `governance_model_configs`. The HTTP API surface is unchanged; the migration handles existing data automatically.

## Security considerations

Budget and rate-limit rows are reused (not duplicated) during migration. The rollback path restores FK references to the provider table and removes the wildcard model config rows, leaving no orphaned governance rows in either direction.

## Checklist

- [x] I read `docs/contributing/README.md` and followed the guidelines
- [x] I added/updated tests where appropriate
- [ ] I updated documentation where needed
- [x] I verified builds succeed (Go and UI)
- [ ] I verified the CI pipeline passes locally if applicable

## Summary by CodeRabbit

* **New Features**
  * Governance policies can be configured for "All Models" on a specific provider, enabling provider-level budget and rate-limit controls.

* **UI Improvements**
  * Model limits table shows "All Models" for wildcard policies.
  * Model picker supports an "All" option for creating provider-scoped policies.
  * Changes now cause related views to refresh so updates appear immediately.

* **Bug Fixes**
  * Provider-scoped governance is applied and cleaned up reliably (including on provider removal).

* **Tests**
  * Added tests covering provider-scoped all-models governance and migrations.

### CODE REVIEW

**Comment by roroghost17** [comment]
* **#3940** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/3940?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#3960](https://github.com/maximhq/bifrost/pull/3960) <a href="https://app.graphite.com/github/pr/maximhq/bifrost/3960?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#3962](https://github.com/maximhq/bifrost/pull/3962) <a href="https://app.graphite.com/github/pr/maximhq/bifrost/3962?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#3939** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/3939?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3938** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/3938?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/maximhq/bifrost/3938?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#3937** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/3937?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `dev`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by roroghost17** [comment]
@coderabbitai full-review

File: framework/configstore/migrations_test.go [hunk 1]
```diff
@@ -2437,3 +2437,43 @@ func TestMigrationAddModelConfigScopeColumns(t *testing.T) {
...
+	assert.Nil(t, prov.RateLimitID, "provider rate_limit_id should be cleared")
+
+	// Idempotency: re-run creates no duplicate wildcard row.
+	require.NoError(t, migrationMigrateProviderGovernanceToModelConfigs(ctx, db))
+	var count int64
+	require.NoError(t, db.Model(&tables.TableModelConfig{}).
+		Where("scope = ? AND model_name = ? AND provider = ?", tables.ModelConfigScopeGlobal, tables.ModelConfigAllModels, "openai").
+		Count(&count).Error)
+	assert.Equal(t, int64(1), count, "re-run must not duplicate the wildcard config")
+}
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply framework/configstore/migrations_test.go:2484]: Acknowledged — the existing test covers the main path well. Edge cases for partial governance (budget-only, rate-limit-only, neither) and multiple providers can be added in a follow-up.

File: plugins/governance/store.go [hunk 2]
```diff
@@ -1224,44 +1278,22 @@ func (gs *LocalGovernanceStore) CheckUserBudget(ctx context.Context, userID stri
...
-	// Uses findModelOnlyConfig for cross-provider model name normalization
-	if mc, configKey := gs.findModelOnlyConfig(ctx, model); mc != nil && mc.RateLimitID != nil {
-		rateLimit := gs.LoadRateLimit(ctx, *mc.RateLimitID)
-		if rateLimit != nil {
-			entityWiseRateLimits[fmt.Sprintf("Model:%s", configKey)] = []*configstoreTables.TableRateLimit{rateLimit}
+		if rateLimit := gs.LoadRateLimit(ctx, *mc.RateLimitID); rateLimit != nil {
+			entityWiseRateLimits[modelConfigEntityKey(mc)] = []*configstoreTables.TableRateLimit{rateLimit}
 		}
 	}
 	return gs.CheckRateLimit(ctx, entityWiseRateLimits, tokensBaselines, requestsBaselines)
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply plugins/governance/store.go:1293]: Each config returned by collectModelConfigsFor has a distinct (model_name, provider, scope, scope_id) tuple enforced by the DB unique index, so modelConfigEntityKey produces a unique key per config and the assignment never overwrites a live entry. The CheckRateLimit early-exit nondeterminism across entities is a pre-existing behavior across all callers — addressed in the PR #3937 thread.

File: framework/configstore/migrations.go [hunk 3]
```diff
@@ -3912,6 +3915,104 @@ func migrationAddModelConfigScopeColumns(ctx context.Context, db *gorm.DB) error
...
+				mc := &wildcards[i]
+				// Re-attach the (reused) budget/rate-limit FK rows to the provider row.
+				if err := tx.Model(&tables.TableProvider{}).Where("name = ?", *mc.Provider).
+					Updates(map[string]any{"budget_id": mc.BudgetID, "rate_limit_id": mc.RateLimitID}).Error; err != nil {
+					return fmt.Errorf("failed to restore governance FKs for provider %q: %w", *mc.Provider, err)
+				}
+				// Drop the wildcard model config; its FK rows now live on the provider again.
+				if err := tx.Delete(&tables.TableModelConfig{}, "id = ?", mc.ID).Error; err != nil {
+					return fmt.Errorf("failed to delete wildcard config for provider %q: %w", *mc.Provider, err)
+				}
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply framework/configstore/migrations.go]: Reverted — the original rollback re-parents budget/rate-limit rows back to provider ownership rather than losing them. Data is moved, not destroyed. Keeping the rollback as-is (same reasoning as the VK governance rollback in #3939).

File: framework/configstore/rdb.go [hunk 4]
```diff
@@ -1155,6 +1155,29 @@ func (s *RDBConfigStore) DeleteProvider(ctx context.Context, provider schemas.Mo
...
+			}
+		}
+		if mc.RateLimitID != nil {
+			if err := txDB.WithContext(ctx).Delete(&tables.TableRateLimit{}, "id = ?", *mc.RateLimitID).Error; err != nil {
+				return err
+			}
+		}
+	}
+	if len(providerModelConfigs) > 0 {
+		if err := txDB.WithContext(ctx).Where("provider = ?", string(provider)).Delete(&tables.TableModelConfig{}).Error; err != nil {
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply framework/configstore/rdb.go]: Fixed in PR #3940 (top of stack) — provider model configs are now deleted by snapshotted IDs (id IN ?) rather than a second WHERE provider=? pass, eliminating the race window.

File: framework/configstore/store.go [hunk 5]
```diff
@@ -256,6 +256,7 @@ type ConfigStore interface {
 
 	// Model config CRUD
 	GetModelConfigs(ctx context.Context) ([]tables.TableModelConfig, error)
+	GetProviderGovernanceModelConfigs(ctx context.Context) ([]tables.TableModelConfig, error) // returns the wildcard "all models on a provider"
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply framework/configstore/store.go]: Fixed in PR #3940 (top of stack) — doc comment now explicitly states the method returns both provider-specific wildcards (provider IS NOT NULL) and the all-providers wildcard (provider IS NULL).

File: plugins/governance/store.go [hunk 6]
```diff
@@ -1069,43 +1071,95 @@ func (gs *LocalGovernanceStore) findScopedModelOnlyConfig(ctx context.Context, s
...
+		add(loadKey(model, provider)) // tier 1: exact model + provider
+	}
+	if mc, _ := gs.findScopedModelOnlyConfig(ctx, scope, scopeID, model); mc != nil {
+		add(mc) // tier 2: exact model, all providers (normalized)
+	}
+	if provider != nil {
+		add(loadKey(modelConfigWildcard, provider)) // tier 3: all models on this provider
+	}
+	add(loadKey(modelConfigWildcard, nil)) // tier 4: all models, all providers
+	return out
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply plugins/governance/store.go:1120]: Fixed in PR #3940 (top of stack) — GetBudgetAndRateLimitStatus now uses collectModelConfigsFor for the model-level section, which covers all four tiers including provider-level wildcard configs.

File: transports/bifrost-http/handlers/governance.go [hunk 7]
```diff
@@ -3093,28 +3130,23 @@ func (h *GovernanceHandler) updateProviderGovernance(ctx *fasthttp.RequestCtx) {
...
-				if provider.RateLimitID != nil {
-					rateLimitIDToDelete = *provider.RateLimitID
-					provider.RateLimitID = nil
-					provider.RateLimit = nil
+			if req.RateLimit.TokenMaxLimit == nil && req.RateLimit.RequestMaxLimit == nil {
+				if mc.RateLimitID != nil {
+					rateLimitIDToDelete = *mc.RateLimitID
+					mc.RateLimitID = nil
+					mc.RateLimit = nil
 				}
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply transports/bifrost-http/handlers/governance.go]: Fixed in PR #3940 (top of stack) — swapped to isRateLimitRemovalRequest.

File: ui/lib/store/apis/governanceApi.ts [hunk 8]
```diff
@@ -545,6 +547,7 @@ export const governanceApi = baseApi.injectEndpoints({
 				method: "PUT",
 				body: data,
 			}),
+			invalidatesTags: ["ProviderGovernance"],
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply ui/lib/store/apis/governanceApi.ts:550]: Fixed in PR #3940 (top of stack) — added the explanatory comment above invalidatesTags for updateModelConfig to match the create and delete endpoints.

File: framework/configstore/rdb.go [hunk 9]
```diff
@@ -1155,6 +1155,33 @@ func (s *RDBConfigStore) DeleteProvider(ctx context.Context, provider schemas.Mo
...
+				return err
+			}
+		}
+	}
+	if len(providerModelConfigs) > 0 {
+		var mcIDs []string
+		if err := txDB.WithContext(ctx).Where("id IN ?", mcIDs).Delete(&tables.TableModelConfig{}).Error; err != nil {
+			return err
+		}
+	}
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply framework/configstore/rdb.go:1183]: Already fixed — mcIDs is populated at line 1171 (mcIDs = append(mcIDs, providerModelConfigs[i].ID)) before the WHERE id IN ? delete. The comment was based on an intermediate diff version.

**Comment by akshaydeo** [comment]
### Merge activity

* **Jun 2, 1:06 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/maximhq/bifrost/3938).
* **Jun 2, 1:09 PM UTC**: [Graphite](https://app.graphite.com/github/pr/maximhq/bifrost/3938) rebased this pull request as part of a merge.
* **Jun 2, 1:10 PM UTC**: @akshaydeo merged this pull request with [Graphite](https://app.graphite.com/github/pr/maximhq/bifrost/3938).
